### PR TITLE
Deleting resource group implemented which also deletes all locks with…

### LINKED ISSUE #28: Project Delete API and Orchestrations

### PR DESCRIPTION
**Pull request by madenwala**
…in the RG. This closes #28

### CODE REVIEW

File: src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            try
+            {
+                var azure = azureSessionService.CreateSession(azureResourceGroup.SubscriptionId);
+                var locks = await azure.ManagementLocks.ListByResourceGroupAsync(azureResourceGroup.ResourceGroupName, true).ConfigureAwait(false);
+                var ids = locks.Select(s => s.Id).ToArray();
+                if (ids.Length > 0)
+                {
+                    await azure.ManagementLocks.DeleteByIdsAsync(ids).ConfigureAwait(false);
+                    // Wait one minute after deleting locks to ensure that Azure has updated itself before trying to delete the RG
+                    await Task.Delay(60 * 1000).ConfigureAwait(false);
```
Comment (markusheiliger) [inline src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs]: no way - we can't rely on a delay here. what if we have to clean up 100 locks. does this take longer?

**Comment by madenwala** [comment]
No. It would be 1 min AFTER all the locks are cleaned up. So no difference if it's 1 lock or 100. I'm not married to this. In my testing I saw this issue but if I waited a few secs it worked. There needs be to a slight delay to ensure that keys are cleared and the delete RG can doesn't throw an exception for a lock that was actually already deletes. Let me know if you have thoughts on other ways to solve this.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by madenwala** [comment]
@markusheiliger I've updated the logic per the discussion on this topic.

File: src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+            try
+            {
+                var azure = azureSessionService.CreateSession(azureResourceGroup.SubscriptionId);
+                await DeleteResourceGroupLocksAsync(azureResourceGroup, azure).ConfigureAwait(false);
+                await azure.ResourceGroups
+                    .DeleteByNameAsync(azureResourceGroup.ResourceGroupName)
+                    .ConfigureAwait(false);
+            }
+            catch (CloudException ex) when (ex.Message.Contains("could not be found", StringComparison.InvariantCultureIgnoreCase))
```
Comment (colbylwilliams) [inline src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs]: The [`CloudException`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.azure.cloudexception?view=azure-dotnet)'s [`Body`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.azure.cloudexception.body?view=azure-dotnet#Microsoft_Rest_Azure_CloudException_Body) property is of type [`CloudError`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.azure.clouderror?view=azure-dotnet).  You should be able to check the [`CloudError.Code`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.rest.azure.clouderror.code?view=azure-dotnet#Microsoft_Rest_Azure_CloudError_Code) instead.

File: src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                await DeleteResourceGroupLocksAsync(azureResourceGroup, azure).ConfigureAwait(false);
+                await azure.ResourceGroups
+                    .DeleteByNameAsync(azureResourceGroup.ResourceGroupName)
+                    .ConfigureAwait(false);
+            }
+            catch (CloudException ex) when (ex.Message.Contains("could not be found", StringComparison.InvariantCultureIgnoreCase)) 
+            // TODO: Is this ok way to check for an exception? This only happens when the RG is already deleted and may not happen often. 
+            // But ignoring this error could prevent the project in the DB from being deleted
+            {
+                Debug.WriteLine("Resource group is already deleted, ignore this exception: " + ex.ToString());
```
Comment (colbylwilliams) [inline src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs]: Please don't use `Debug.WriteLine`.  Inject an `Microsoft.Extensions.Logging.Ilogger` into the function as a param and log using the `Ilogger`

File: src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            }
+            catch (Exception ex)
+            {
+                Debug.WriteLine(ex.ToString());
+                throw;
+            }
+        }
+
+        private async Task DeleteResourceGroupLocksAsync(AzureResourceGroup azureResourceGroup, IAzure azure)
```
Comment (colbylwilliams) [inline src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs]: @markusheiliger please review

Comment (madenwala) [reply src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs]: Fixed in latest push.

File: src/TeamCloud.Orchestrator/Orchestrations/Projects/ProjectDeleteOrchestration.cs [hunk 5]
```diff
@@ -43,8 +44,9 @@ await functionContext
             var providerCommands = teamCloud.Providers.Select(provider => new ProviderCommand { Command = command, Provider = provider });
             var providerCommandTasks = providerCommands.Reverse().Select(providerCommand => functionContext.CallSubOrchestratorAsync<ProviderCommandResult>(nameof(ProviderCommandOrchestration), providerCommand));
 
-            var providerCommandResults = await Task
```
Comment (madenwala) [inline src/TeamCloud.Orchestrator/Orchestrations/Projects/ProjectDeleteOrchestration.cs]: This line was delete in merge, was added back in latest push.

File: src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs (hunk 2 above)
Comment (madenwala) [reply src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs]: Fixed in latest push.

File: src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs (hunk 1 above)
Comment (madenwala) [reply src/TeamCloud.Orchestrator/Orchestrations/Azure/AzureResourceGroupDeleteActivity.cs]: This was replaced with a loop.  Loop will abort after 10 tries to make sure it's not ever permanently stuck in a loop.

**Pull request merged**